src/runtime-rs/crates/runtimes/virt_container/src/sandbox.rs

```diff
@@ -223,6 +227,43 @@ impl VirtSandbox {
         })
     }
 
+    async fn prepare_vm_socket_config(&self) -> Result<ResourceConfig> {
+        // get vm agent socket
+        // There're two kinds of vm sockets:
+        // - Hybrid Vsock: hvsock://<path>:<port>.
+        // -- Example: "hvsock:///x/y/z/kata.hvsock:port"
+        // - Vsock: vsock://<cid>:<port>
+        // -- Example: "vsock://31513974:1024"
+        let agent_socket = self
+            .hypervisor
+            .get_agent_socket()
+            .await
+            .context("get agent socket")?;
+
+        let uri_fields: Vec<&str> = agent_socket.split("://").collect();
+        if uri_fields.len() != 2 {
+            return Err(anyhow!("invalid URI"));
+        }
+
+        // Generate device configs for vm sockets depending on the URI scheme.
+        let uri_scheme = uri_fields[0].to_uppercase();
```

Comment:
It might make sense to define the values of `SCHEME_VSOCK` and `SCHEME_HYBRID_VSOCK` in lowercase in the first place.  Then there would be no need to uppercase the scheme here, right?